### PR TITLE
feat(vue-app): support `app/router.scrollBehavior.js` and deprecate `scrollBehavior`

### PR DESCRIPTION
**Pull request by atinux**
This PR allows to overwrite application and modules templates with the `app/` directory directly.

## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Not really a feature, but a refactor to move the router `scrollBehavior` to `router.scrollBehavior.js`.

**Why?**

In order to stop serializing function and start opening another way of changing the scrollBehavior with a file.
It also avoid restarting the whole Nuxt application since it's actually a config related to the Vue app.

**Before:**

```js
// nuxt.config.js
export default {
  router: {
    scrollBehavior: function (to, from, savedPosition) {
       return { x: 0, y: 0 }
    }
  }
}
```

**After:**

```js
// app/router.scrollBehavior.js
export default function(to, from, savedPosition) {
  return { x: 0, y: 0 }
}
```

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly. (PR: https://github.com/nuxt/docs/pull/1472)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by galvez** [comment]
YES, thank you -- I was getting close to going down the `acorn` rabbit hole just to write us a working serializeFunction. It's definitely much better to have these imported as live JavaScript instead.

**Comment by Atinux** [comment]
I plan to do the same logic to most of the configuration option that have to be serialised (head, transition, etc) 😄

File: packages/config/src/options.js [hunk 1]
```diff
@@ -35,6 +35,12 @@ export function getNuxtConfig (_options) {
     options._routerBaseSpecified = true
   }
 
+  // TODO: Remove for Nuxt 3
+  // router.scrollBehavior -> app/router.scrollBehavior.js
+  if (options.router && typeof options.router !== 'undefined') {
+    consola.warn('`router.scrollBehavior` property is deprecated in favor of using `~/app/router.scrollBehavior.js` file')
```
Comment (pi0) [inline packages/config/src/options.js]: Can we probably guide users to the new docs for quick migration?

**Review by pi0 — DISMISSED** [review]
Neat :+1:

File: packages/config/src/options.js (hunk 1 above)
Comment (Atinux) [reply packages/config/src/options.js]: Yes,  I will make a PR to the docs + link in the warn for it :)

**Comment by Atinux** [comment]
I need to add support for watching `app/**/*.(js|vue)` when creating or removing file there.

**Comment by Atinux** [comment]
Watching `app/` is done, I also added `dir.app` key in config

**Comment by Atinux** [comment]
We can now overwrites custom templates given in `build.templates` with `app/${dst}`.

Example of a module adding `press/common/pages/source.vue` with `this.addTemplate()`, the user could overwrite it by adding `app/common/pages/source.vue`.

Nuxt will automatically watch the `app/` dir and handle file added/updated/removed for it so the module author won't have to deal with watchers.

**Review by galvez — DISMISSED** [review]

**Review by pimlie — COMMENTED** [review]
Is `app` discriminatory enough to be used for overwriting the nuxt-app / templates? Its such a common name that I can imagine people are already using it in their projects.

**Comment by Atinux** [comment]
@pimlie actually it was already doing this since a long time (can't remember the version). Actually i found the name perfect to overwrite our default `vue-app` template (and also modules like NuxtPress: `app/press/common/pages/source.vue`)

**Review by pimlie — DISMISSED** [review]
Ok, thanks :)

File: packages/config/src/options.js [hunk 2]
```diff
@@ -35,6 +35,12 @@ export function getNuxtConfig (_options) {
     options._routerBaseSpecified = true
   }
 
+  // TODO: Remove for Nuxt 3
+  // router.scrollBehavior -> app/router.scrollBehavior.js
+  if (options.router && typeof options.router !== 'undefined') {
```
Comment (kevinmarrec) [inline packages/config/src/options.js]: Seems the warning would display whenever `router` is defined, I think it should check if `scrollBehavior` is defined too.

**Review by kevinmarrec — CHANGES_REQUESTED** [review]
Great work ! 

See my comment for the little requested change.

Aside that, I think we may need a TypeScript typing change around https://github.com/nuxt/typescript/blob/master/packages/types/config/router.d.ts that may exclude `scrollBehavior` from `RouterOptions` so that people doesn't have autocompletion on obsolete option.

It also made me notice that we don't have configuration types for `dir` 🤦🏼😁

File: packages/config/src/options.js (hunk 2 above)
Comment (Atinux) [reply packages/config/src/options.js]: Fixed, good catch @kevinmarrec

**Comment by Atinux** [comment]
Good catch for typings Kevin, I think it's fine if they have autocomplete and then Nuxt warning 🤔

**Review by pi0 — APPROVED** [review]

File: packages/builder/src/builder.js [hunk 3]
```diff
@@ -405,42 +405,47 @@ export default class Builder {
...
+        typeof t === 'object' ? t : undefined
+      )
+    })
     const customTemplateFiles = this.options.build.templates.map(
       t => t.dst || path.basename(t.src || t)
     )
+    const templatePaths = uniq([
+      // Modules & user provided templates
+      // first custom to keep their index
+      ...customTemplateFiles,
```
Comment (pi0) [inline packages/builder/src/builder.js]: Shouldn't they come second? As they should override built-ins? :thinking:

**Review by kevinmarrec — APPROVED** [review]

**Comment by pi0** [comment]
2.9.0 notes updated to mention

**Comment by manniL** [comment]
I really like the fact that we avoid serialization with that step, however (as briefly discussed with @Atinux), this means we can't provide data from the `nuxt.config.js` inside such a external file in `app` (at the moment at least). For `scrollBevahior`, this might not have a huge impact, however for further split ups of things like transitions or `head` it could create a few problems if we don't handle it. :relaxed:

**Pull request merged**